File: src/buffer_op/gpu/fade.c
__kernel void fade_kernel(
    __global uint* buf,
    uint width,
    uint height,
    uint color_operand,
    int op_code, // 0 = Add, 1 = Subtract
    int use_alpha
) {
    uint gid = get_global_id(0);
    if (gid >= width * height) return;

    uint px = buf[gid];

    uint r1 = (px >> 24) & 0xFF;
    uint g1 = (px >> 16) & 0xFF;
    uint b1 = (px >> 8) & 0xFF;
    uint a1 = px & 0xFF;

    uint r2 = (color_operand >> 24) & 0xFF;
    uint g2 = (color_operand >> 16) & 0xFF;
    uint b2 = (color_operand >> 8) & 0xFF;
    uint a2 = color_operand & 0xFF;

    int r_sum = (int)r1 + (int)r2;
    int g_sum = (int)g1 + (int)g2;
    int b_sum = (int)b1 + (int)b2;
    int a_sum = (int)a1 + (int)a2;

    int r_diff = (int)r1 - (int)r2;
    int g_diff = (int)g1 - (int)g2;
    int b_diff = (int)b1 - (int)b2;
    int a_diff = (int)a1 - (int)a2;

    int rf, gf, bf, af;

    if (op_code == 0) { // Add
        rf = r_sum > 255 ? 255 : r_sum;
        gf = g_sum > 255 ? 255 : g_sum;
        bf = b_sum > 255 ? 255 : b_sum;
        af = use_alpha ? (a_sum > 255 ? 255 : a_sum) : a1;
    } else { // Subtract
        rf = r_diff < 0 ? 0 : r_diff;
        gf = g_diff < 0 ? 0 : g_diff;
        bf = b_diff < 0 ? 0 : b_diff;
        af = use_alpha ? (a_diff < 0 ? 0 : a_diff) : a1;
    }

    buf[gid] = ((uint)rf << 24) | ((uint)gf << 16) | ((uint)bf << 8) | ((uint)af);
}

File: src/buffer_op/gpu/fade.rs
use crate::core::context::gpu::GpuContext;
use crate::utils::color::math::ColorOperation;


use crate::buffer_op::gpu::kernel_bundle::KernelBundle;
#[cfg(feature = "gpu")]
use ocl::Kernel;

/// GPU-accelerated fade effect using OpenCL.
/// Applies a color operation to each pixel in the buffer.
///
/// # Parameters
/// * `cpu_buf` - The mutable CPU-side buffer (RGBA pixels packed in u32)
/// * `width` - Width of the image
/// * `height` - Height of the image
/// * `color_operand` - The color operand (as RGBA packed in u32)
/// * `op` - Color operation (Add or Subtract)
/// * `use_alpha` - Whether alpha is affected
/// * `gpu_context` - The GPU context for managing OpenCL resources
#[allow(dead_code)]
pub fn fade_gpu(
    cpu_buf: &mut [u32],
    width: u32,
    height: u32,
    color_operand: u32,
    op: ColorOperation,
    use_alpha: bool,
    gpu_context: &mut GpuContext,
) -> Result<(), String> {
    #[cfg(feature = "gpu")]
    {
        let buf_len = (width * height) as usize;
        let bundle = fade_get_kernel(
            cpu_buf,
            width,
            height,
            color_operand,
            op,
            use_alpha,
            gpu_context,
        )?;
        let kernel = bundle.kernel;
        let gpu_buf = bundle.buffers.get(0).ok_or("No GPU buffer")?;

        gpu_buf
            .write(cpu_buf.as_ref())
            .enq()
            .map_err(|e| format!("Upload error: {e}"))?;
        unsafe {
            kernel
                .enq()
                .map_err(|e| format!("Kernel enqueue error: {e}"))?;
        }
        gpu_buf
            .read(cpu_buf)
            .enq()
            .map_err(|e| format!("Download error: {e}"))?;
        Ok(())
    }
    #[cfg(not(feature = "gpu"))]
    {
        let _ = (
            cpu_buf,
            width,
            height,
            color_operand,
            op,
            use_alpha,
            gpu_context,
        );
        Err("GPU support is not enabled.".to_string())
    }
}


pub fn fade_get_kernel(
    cpu_buf: &mut [u32],
    width: u32,
    height: u32,
    color_operand: u32,
    op: ColorOperation,
    use_alpha: bool,
    gpu_context: &mut GpuContext,
) -> Result<KernelBundle, String> {
    #[cfg(feature = "gpu")]
    {
        let kernel_src = include_str!("fade.c");
        let kernel_name = "fade_kernel";
        let program_name = "fade_program";
        let buf_len = (width * height) as usize;

        if gpu_context.get_program(program_name).is_none() {
            gpu_context
                .load_program(kernel_src, program_name)
                .map_err(|e| format!("Program load error: {e}"))?;
        }

        let program = gpu_context
            .get_program(program_name)
            .ok_or("Program missing unexpectedly")?;
        let queue = gpu_context
            .queue
            .as_ref()
            .ok_or("Missing GPU queue")?
            .clone();
        let gpu_buf = gpu_context.get_or_create_buffer(buf_len, queue.as_ref(), cpu_buf)?;
        let queue_value = queue.as_ref().clone();

        let kernel = Kernel::builder()
            .program(&program)
            .name(kernel_name)
            .queue(queue_value)
            .global_work_size(buf_len)
            .arg(&gpu_buf)
            .arg(width)
            .arg(height)
            .arg(color_operand)
            .arg(op as i32)
            .arg(use_alpha as i32)
            .build()
            .map_err(|e| format!("Kernel build error: {e}"))?;

        Ok(KernelBundle::new(kernel, vec![gpu_buf]))
    }
    #[cfg(not(feature = "gpu"))]
    {
        let _ = (cpu_buf, width, height, color_operand, op, use_alpha, gpu_context);
        Err("GPU support is not enabled.".to_string())
    }
}
